src/catalog_xliff.h

```diff
@@ -89,6 +89,7 @@ class XLIFFCatalog : public Catalog
     bool HasCapability(Cap cap) const override;
 
     static bool CanLoadFile(const wxString& extension);
+    static const wxString& GetBaseDir();
```

Comment:
Moved+renamed to `CloudSyncDestination::GetCacheDir()` which returns common cache dir (i.e. without `Crowdin` subdir at the end), while still introduced  `XLIFFCatalog::GetCacheDir()`  https://github.com/vslavik/poedit/blob/7239fc0c686f7c995d9f0eec623d14d39e9ead56/src/catalog_xliff.h#L93 as well since it adds Crowdin-specific subpath. I understand that it more appropriate place for it would be in `CrowdinSyncDestination` by same analogy with base one but the problem that `CrowdinSyncDestination` is in `GUI` part of code on macOS while it used from `Non-GUI` part https://github.com/vslavik/poedit/blob/7239fc0c686f7c995d9f0eec623d14d39e9ead56/src/catalog_xliff.cpp#L604 . But if `CloudSyncDestination::GetCacheDir()` is not acceptable for you at all by design - I can move `CrowdinSyncDestination` to some new separate `Non-GUI` file.

Reply:
I still don't understand why do you need `GetCacheDir` in the file object. Poedit on `master` doesn't have it, yet is able to fetch files from Crowdin and edit them. Clearly it is not *necessary*, but you think it is needed/useful/something. I don't understand why.

Putting cache stuff into a class that handles file format doesn't really make sense to me, and putting *Crowdin*-specific stuff there doesn't either.

`CloudSyncDestination` is non-GUI class, and `CrowdinSyncDestination` is in `crowdin_gui.*` files by… my laziness, I suppose, as it has no GUI code in it. It can be trivially moved to `crowdin_client.*` files, cannot it?

Reply:
> I still don't understand why do you need GetCacheDir in the file object

If you about using it here https://github.com/vslavik/poedit/blob/7239fc0c686f7c995d9f0eec623d14d39e9ead56/src/catalog_xliff.cpp#L604 (which mentioned above) - it's just shortest way I was able to find to extract Crowdin project and file IDs from file path to where they are injected from here https://github.com/vslavik/poedit/blob/7239fc0c686f7c995d9f0eec623d14d39e9ead56/src/crowdin_gui.cpp#L491. What in turn again shortest way I was able to find to allow `Xliff` files that opened from local cache to be saved/synced back to Crowdin without keeping above IDs inside Xliff file itself as some custom attributes or metadata. What in turn would be most straightforward way to reach above feature of course but as discussed with Crowdin team - may lead to unexpected result for customers workflows relying on diffing/hashing files like git or other version controls, i.e. it will appear as unexpected/unintended change of file content in those workflows. I.e. regardless if above IDs will be introduced on server side by Crowdin or by Poedit while just downloading that file from Crowdin. In both cases those IDs will be saved (uploaded) back to Crowdin from where will already come anywhere else in customers workflows.

Reply:
> . What in turn again shortest way I was able to find 

Seems like a really convoluted way to replace `m_fileName.GetName()` to me, but I may be missing some subtlety...

> to allow Xliff files that opened from local cache to be saved/synced back to Crowdin

I don't think such capability is a requirement at all right now — that it works now is just a side-effect.

I suggest we focus on the normal flow now (i.e. open with the Crowdin button, edit, save), and revisit persistent caching afterwards.

However I still don't see how having cache directory in XLIFF-only code facilitates that. For detection of a locally saved file without embedded metadata, `IsFromCrowdin` could be extended to look at the filename too (not just for XLIFF; for all).

> What in turn would be most straightforward way to reach above feature of course, but as discussed with Crowdin team - may lead to unexpected results for customers workflows relying on diffing/hashing of files like git or other version controls, i.e. it will appear as unexpected/unintended change of file content in those workflows. And this is regardless if above IDs will be introduced on server side by Crowdin or by Poedit on downloading from Crowdin.

Poedit certainly shouldn't add such metadata itself. Exported XLIFF files from Crowdin certainly should have them (at least a proper header). I'm not sure why one-time addition of them after export when they improve the code would be a VCS issue, if I understand you correctly, or why would Crowdin say that given that they change tiny details in exported files *all the time* and just recently [broke PO files validation](https://github.com/vslavik/poedit/commit/76b4835c6df6c6b6613d2ec81a2b6827546e6c1d) without much of a concern...

Reply:
> Seems like a really convoluted way to replace m_fileName.GetName() to me, but I may be missing some subtlety...

Could you please point me on exact file+line of code you mean so I would make sure I understand you correctly? Or you just about inserting project and file ID into filename itself? If you about this - it actually would allow customer to even move file downloaded from Crowdin to some totally different location outside of the cache but just preserving it name. And then it can be saved/synced back to Crowdin on next open.

>> to allow Xliff files that opened from local cache to be saved/synced back to Crowdin

> I don't think such capability is a requirement at all right now — that it works now is just a side-effect.

> I suggest we focus on the normal flow now (i.e. open with the Crowdin button, edit, save), and revisit persistent caching afterwards.

Sorry, but I don't understand what you mean especially  "and revisit persistent caching afterwards.". Since without that implementation - when user did "Open from Crowdin", editing, "Save" or "Sync", close Poedit, then open same file from Resent and repeat

Reply:
`CloudSyncDestination` is non-GUI class, and `CrowdinSyncDestination` is in crowdin_gui.* files by… my laziness, I suppose, as it has no GUI code in it. It can be trivially moved to crowdin_client.* files, cannot it?

> However I still don't see how having cache directory in XLIFF-only code facilitates that.

Now I seems fully got what you mean. It's really already my laziness to use `GetCacheDir()` there https://github.com/vslavik/poedit/blob/7239fc0c686f7c995d9f0eec623d14d39e9ead56/src/catalog_xliff.cpp#L604 just to get length of cache dir path string as `GetCacheDir().size()`. So now I fully see your alternative to move `GetCacheDir()` from `XLIFFCatalog` to `CrowdinSyncDestination` and to extract project and file IDs from filename **without it** here https://github.com/vslavik/poedit/blob/7239fc0c686f7c995d9f0eec623d14d39e9ead56/src/catalog_xliff.cpp#L604 so even no necessity to move  `CrowdinSyncDestination` to `crowdin_client.*` in this case

Reply:
> Could you please point me on exact file+line of code you mean so I would make sure I understand you correctly?

Sorry. You referenced this code as being the reason:
```
m_fileName.Right(m_fileName.size() - GetCacheDir().size() - 1).ToLong(&m_projectId); 
```
That can be done in a simpler (*and* more robust) way as
```
m_fileName.GetName().ToLong(&m_projectId); 
```

And in any case, none of that needs to be done in XLIFFCatalog, that's what (format-independent) `IsFromCrowdin` is for. 

> Or you just about inserting project and file ID into filename itself? If you about this - it actually would allow customer to even move file 

I didn't want to add more noise, so didn't mention this, but yes, that's along the lines I would do it.

> Since without that implementation - below would not work:

Right. I don't necessarily *expect* this to work, as I said, that's a side-effect: in step 5, I'd expect normal flow to be "Open from Crowdin" again, so not putting Crowdin-opened files into Recent Files would be OK too.

However, if all that is really needed to keep the sync is having the information embedded in the full path under the cache dir, then I think that indeed putting it into the filename only (e.g. as CrowdinSynced.PROJECT.FILENAME.xliff or something like that?), and moving the detection into IsFromCrowdin, is the best course of action.

Reply:
> For detection of a locally saved file without embedded metadata, `IsFromCrowdin` could be extended to look at the filename too (not just for XLIFF; for all)

Really makes sense. It's more universal, reusable and arranged/structured. Will re-arrange in this way.

Reply:
> Right. I don't necessarily expect this to work, as I said, that's a side-effect: in step 5, I'd expect normal flow to be "Open from Crowdin" again, so not putting Crowdin-opened files into Recent Files would be OK too.

From the point of Crowdin team - it appears as useful feature for their customers :) - i.e. to be able to reopen Crowdin's file from Recent and then to Save/Sync it back after editing. For `IsFromCrowdin` fully makes sense, agreed and will re-arrange. Thanks for pointing me out!

Reply:
>  That can be done in a simpler (and more robust) way as
> `m_fileName.GetName().ToLong(&m_projectId); `

Great hint, thanks a lot! Will re-arrange

Reply:
> Exported XLIFF files from Crowdin certainly should have them (at least a proper header). I'm not sure why one-time addition of them after export when they improve the code would be a VCS issue, if I understand you correctly.

Yes, you understand me correctly, at least how I understand what Crowdin team answered/explained me. So I just guessed they better than me knows possible workflows of their customers using Crowdin directly, with Poedit or with some another than Poedit tools

> or why would Crowdin say that given that they change tiny details in exported files all the time and just recently broke PO files validation without much of a concern...

It might worth for you to talk to them about this in our email discussion to get exhaustive answer

Reply:
Done everything discussed to original 3c977c95d84823ef51b3d5fd0b56140128ce2269 and force-pushed.

Marking as resolved or please let me know if something is missed or still needs corrections. Thanks a lot for all advices and hints here!